### ISSUE TITLE
Wildcard in `merge`

### ISSUE THREAD

**Issue by Nadrieril** [dhall-lang/dhall-lang#961]
It has come up several times already that having to list all alternatives in `merge` is sometimes painful. @Gabriel439 [suggests](https://discourse.dhall-lang.org/t/sum-type-value-equal-predicate/185/11) we could support a special wildcard field in the merge record that would act as a wildcard, e.g.

```dhall
let Color = < Red | Green | Blue >
let isRed = \(c: Color) -> merge {
		        Red = True,
				_ = False,
		    } c
```

This would make `merge` closer to classic pattern-matching, and less boilerplate-y in many cases.

**Comment by Gabriel439** [comment]
... and also to reiterate here, this would likely also entail forbidding `_` as an alternative name

**Comment by SiriusStarr** [comment]
> ... and also to reiterate here, this would likely also entail forbidding `_` as an alternative name

This presumably further encourages `_` for use with #74.  If it's forbidden from unions, it seems a natural step for records as well.  (Though we still have the normalization issue?)

**Comment by TristanCacqueray** [comment]
Unless someone wants to take this over, I'd be happy to do the implementation. The initial attempt proposed in https://github.com/dhall-lang/dhall-haskell/pull/1751 seems to work:

```
./dist/build/dhall/dhall <<EOF
let Color = < Red | Green | Blue >
let isRed = \(c: Color) -> merge {Red = True, _ = False} c
in [isRed Color.Red, isRed Color.Green]
EOF

[ True, False ]
```

**Comment by sjakobi** [comment]
This looks similar to catch-alls in Haskell's case expressions:

```haskell
data Color = Red | Green | Blue

isRed :: Color -> Bool
isRed color = case color of
  Red -> True
  _ -> False
```

The usual problem with catch-alls is that programmers forget to update these case expressions when they change the type of the scrutinee, for example by adding a constructor. Similarly to the `...` pattern discussed in #74, a catch-all allows the programmer to opt out of the exhaustiveness check which would otherwise remind them to update their `merge` or `case` expression.

A safe alternative to catch-alls is called or-patterns. Rust AFAIK has them. For Haskell there's a [proposal](https://github.com/osa1/ghc-proposals/blob/or_patterns/proposals/0000-or-patterns.rst). For the example above it would look like this:

 ```haskell
isRed :: Color -> Bool
isRed color = case color of
  Red -> True
  (Green; Blue) -> False
```

The benefit is that programmers don't have to repeat the common alternative (`False` in the example above) for each constructor, while retaining the safety of the exhaustiveness check.

Now I'm wondering how to best translate or-patterns into Dhall. It would be convenient to have some syntactic sugar to construct multiple record entries from several fields and a single value…

How about something like this?

```dhall
let isRed = \(c: Color) -> merge { Red = True, [ Green, Blue ] = False } c
```

**Comment by Gabriel439** [comment]
@sjakobi: It's not clear that your proposed alternative would fix the problem of exhaustiveness checking.  Using the example of colors, the only thing that would change is that the user would define a record containing all the colors set to false:

```dhall
let default = { [ Red, Green, Blue ] = False }

in  isRed = \(color : Color) -> merge (default // { Red = True }) c
```

... and that would be just as vulnerable to downstream functions not being audited when the user extended the `default` record.

**Comment by sjakobi** [comment]
@Gabriel439 The "default handler pattern" you describe is already possible today. IMHO it's less convenient than wildcards/catch-alls though, which, in the context of escape hatches for safety checks, is arguably a good thing.

I also suspect that the default handler pattern is only truly convenient for simple predicate-type `merge`s. Wildcards would be an alluring escape hatch in all kinds of `merge`s.

**Comment by Gabriel439** [comment]
@sjakobi: To be clear: I support the proposal of wildcard handlers instead of creating a record of default handlers.

What I meant to say was that I don't think the preference for exhaustiveness checking is an argument against wildcard pattern matches because even if we don't standardize wildcard matches people will bypass exhaustiveness checking anyway using records of default handlers.

**Comment by sjakobi** [comment]
What I don't understand so far is what we gain by making it more convenient to opt out of the exhaustiveness check.

**Comment by TristanCacqueray** [comment]
@sjakobi that enables removing boilerplate to answer the problem described in https://discourse.dhall-lang.org/t/sum-type-value-equal-predicate/185/10

**Comment by sjakobi** [comment]
@TristanCacqueray Thanks for the link!

Is this mainly about union equality then? My impression is that the amount of boilerplate saved by using wildcards is only a very small part of the total boilerplate necessary to define equality for e.g. `Color`:

<details><summary>This is what you might currently write</summary>

```dhall
let Color = < Red | Green | Blue >

let defaultHandlers = { Red = False, Green = False, Blue = False }

let isRed = λ(x : Color) → merge (defaultHandlers with Red = True) x

let isGreen = λ(x : Color) → merge (defaultHandlers with Green = True) x

let isBlue = λ(x : Color) → merge (defaultHandlers with Blue = True) x

let eqColor =
        λ(x : Color)
      → λ(y : Color)
      → merge { Red = isRed y, Green = isGreen y, Blue = isBlue y } x

let example = assert : eqColor Color.Red Color.Red ≡ True

in  eqColor
```

</details>

<details><summary>This is what you could write with wildcards</summary>

```dhall
let Color = < Red | Green | Blue >

let isRed = λ(x : Color) → merge { Red = True, _ = False } x

let isGreen = λ(x : Color) → merge { Green = True, _ = False } x

let isBlue = λ(x : Color) → merge { Blue = True, _ = False } x

let eqColor =
        λ(x : Color)
      → λ(y : Color)
      → merge { Red = isRed y, Green = isGreen y, Blue = isBlue y } x

let example = assert : eqColor Color.Red Color.Red ≡ True

in  eqColor
```

</details>

The main benefit is that you don't have to define `defaultHandlers` (which could be defined more compactly with the syntactic sugar I suggested above), and that each `isX` predicate gets _slightly_ shorter.

Is this enough convenience to warrant adding such a versatile footgun? I'm not yet convinced…

**Comment by Gabriel439** [comment]
@sjakobi: The wildcard approach is not much simpler if the user wants to write an equality function, but in many of these cases the equality function wasn't what the user cared about anyway.  In many of these cases the equality function was a circuitous way of just trying to select one specific alternative from a list.

For example, the user might only want to filter out all elements matching a specific alternative, like this:

```dhall
Prelude.List.filter Foo (\(x : Foo) -> merge { Red = True, _ = False } x.color)
```

... and in such an example the wildcard based approach is more ergonomic.

**Comment by TristanCacqueray** [comment]
I agree the benefit is rather limited with regards to union equality test, though it's one less things to do when updating such union that can be compared. And having a wildcard seems useful in other situation.

Could this wildcard addition be somehow used by a followup union comparator implementation? For example by making `Color.Red == Color.Green` desugar into `merge {Red = True, _ = False} Color.Green`, or would that be un-necessary?

**Comment by sjakobi** [comment]
OK, I concede that there are cases (like predicates) where wildcards seem to be the appropriate tool.

I still worry that we'll see wildcards used a lot in scenarios where opting out of the exhaustiveness check is risky. To provide a slightly less concise but safer alternative, I think we should also standardize something similar to my or-pattern adaptation above.

On a syntactic level, assigning special meaning to the `_` key also seems rather awkward. I suspect that this will be a dead end for pattern matches in `merge` – I don't see how we could fit any more interesting patterns into record keys.

**Comment by TristanCacqueray** [comment]
@sjakobi raised good points and for what it's worth I'd be more than happy to close the PRs I made related to that issue if needed.

**Comment by Nadrieril** [comment]
I think what we'd really want is an actual pattern matching syntax but that's obviously a whole other level of complexity. So I don't mind adding awkwarder features like the wildcard to get closer to that without the full complexity of pattern-matching

**Comment by Gabriel439** [comment]
@sjakobi: I don't believe your or-pattern proposal would solve the original problem, though, because the or-pattern only condenses the code by a constant factor.  The rationale for having a wildcard pattern match is for unions where the number of alternatives is large (e.g. 24) and you only care about one.

My intuition here is that what users want is a way to select out just one alternative, so if there is an alternative approach to do that which doesn't involve wildcard pattern matches that would solve the original problem.  Or, perhaps we could require that all wildcard pattern matches must ignore all but one constructor (e.g. `{ A = x, B = y, _ = z }` is always rejected because it does not ignore two constructors)

**Comment by sjakobi** [comment]
> Or, perhaps we could require that all wildcard pattern matches must ignore all but one constructor (e.g. `{ A = x, B = y, _ = z }` is always rejected because it handles two non-wildcard constructors)

That's an interesting idea! People might work around this (surprising) constraint by combining multiple `merge`s that contain wildcards though, e.g. `isRed c || isBlue c`. But it should help stop people from arbitrarily slapping wildcards on many `merge`s where they'd be better advised to ensure exhaustiveness.

I guess my remaining complaint is about the awkwardness of a special `_` field. Maybe we can retire this syntax once we have proper support for pattern matching on unions…

But if other people like the proposal, I won't attempt to block it any longer. :)

**Comment by Gabriel439** [comment]
@sjakobi: Another option might be to generalize `==`/`!=` to work on enums with empty alternatives.  The basic idea would be that for both operators:

* The left and right side must both be the same enum type

* The alternatives being compared must both by empty (otherwise it is a type error)

  Alternatively: *all* alternatives of the enum type must be empty (not just the ones being compared)

* They are equal if they are the same alternative, unequal otherwise

**Comment by sjakobi** [comment]
> Alternatively: _all_ alternatives of the enum type must be empty (not just the ones being compared)

Wouldn't that be necessary anyway, so you can write something like

```dhall
\(x : MyEnum) -> \(y : MyEnum) -> x == y
```

?

I suspect that once we start with enums with empty alternatives, we'll get requests to support e.g. `< A | B : Bool >`, `<A | B : { n : Natural } >`, records with comparable fields, etc. too.

**Comment by Gabriel439** [comment]
@sjakobi: Oh yeah, good point.  All of the alternatives would need to be empty.

Yeah, the more I think about this the more I think that we actually need is `==` / `!=` for enums.  I think that would also satisfy all of the original use cases presented in the Discourse thread and I believe would mitigate the concerns that @sjakobi has mentioned

@TristanCacqueray @Nadrieril: Would it be alright if we were to standardize that instead?

**Comment by Nadrieril** [comment]
I think both are rather different features and would be both useful. I'm open to `==` for unions but I don't think it entirely replaces the need for wildcards

**Comment by TristanCacqueray** [comment]
@Gabriel439 it would be alright to standardize `==` for enums. Perhaps we could do that in parallel as wildcards may be useful in other contexts?

**Comment by matthew-piziak** [comment]
I was in that Discourse thread. For what it's worth equality and wildcards would each be sufficient for my usecase. Equality would be more explicit of course, while wildcards would be syntactically nicer for users coming from other pattern-matching languages.

**Comment by Gabriel439** [comment]
Yeah, that's a good point that wildcards can handle more cases than equality, so for now I'll prefer the wildcard match over the equality idea

**Comment by arianvp** [comment]
A place where I am having a lot of issues is where I want to define some utility functions for `dhall-kubernetes` which work on the union; but having to write out all 450 cases of the case of statement whilst I only want to match against `Deployment` and `StatefulSet` is extremely painful

**Comment by dominik-steenken** [comment]
Hi, sorry for butting in on a pretty old issue, but is this still being worked on?

**Comment by Gabriel439** [comment]
@dominik-steenken: No, this has stalled

**Comment by dominik-steenken** [comment]
ok. i guess this was already implied by my question, but i would be very interested in a solution / implementation for this, since we have a lot of merging of typed records in our configurations, and this would massively cut down on the boilerplate. So, i just would like to add my vote to the folks who would love to see this in dhall :).

**Comment by imuli** [comment]
I'd also really like to see this, not so much from an ergonomics perspective but from being able to loosen the coupling between Dhall and the data I'm injecting for forward (from the Dhall perspective)/backward (from the data injector perspective) compatibility.

I may be able to inject a function that creates a catchall given a default for every type of data we're injecting, but that's clearly a hack and wildcard merging would be much better!

**Comment by Gabriella439** [comment]
@imuli: Can you give an example of what you're referring to?

**Comment by imuli** [comment]
So imagine we have a description of polyhedra like this:

```haskell
data Base = Tetrahedron | Cube
data Mod = Truncate | Stellate | Excavate
data Polyhedron = Polyhedron { base: Base, mod: Maybe Mod }
```

We inject that into dhall producing types like this:

```dhall
let Base = < Tetrahedron | Cube >
let Mod = < Truncate | Stellate | Excavate >
let Polyhedra = { base: Base, mod: Optional Mod }
```

But different polyhedron makers only know about the polyhedra they can make:

```dhall
let PolyhedraA = < Cube | TruncatedCube | TruncatedTetrahedron >
let PolyhedraB = < ExcavatedCube | Cube | StellatedCube | Tetrahedron | TruncatedTetrahedron >
```

(assume that's partial truncation, so a truncated tetrahedron isn't just a tetrahedron)

So we write the matching function:

```dhall
let matchA = (\poly: Polyhedra) -> merge
  { Tetrahedron = merge
    { None = None PolyhedraA
    , Some = (\mod: Mod) -> merge
      { Truncate = Some PolyhedraA.TruncatedTetrahedron
      , Stellate = None PolyhedraA
      , Excavate = None PolyhedraA
      } mod
    } poly.mod
  , Cube = merge
    { None = Some PolyhedraA.Cube
    , Some = (\mod: Mod) -> merge
      { Truncate = Some PolyhedraA.TruncatedCube
      , Stellate = None PolyhedraA
      , Excavate = None PolyhedraA
      } mod
    } poly.mod
  } poly.base
```

and similar for `matchB`.

But then we add another polyhedron maker, and it also does icosahedra.
We now need to add `Icosahedron` to our `Base` type, which breaks the existing merges for makers A and B.

**Comment by Gabriella439** [comment]
Alright, then I'd approve a change to add language support for wildcard matches, modulo some unforeseen language design issue that hasn't already been noted here.

I do agree with @sjakobi that it raises the potential for writing pattern matches that ignore new cases that are added.  For example, this is an issue we have at work (in Haskell) where we have some large pattern matches where we ban the use of wildcard pattern matches to ensure that all use cases are handled.

Maybe a possible compromise would be to add an `exhaustive merge` variation on `merge` that bans the use of `_` within that `merge`, sort of like language-checked documentation that a pattern match shouldn't use a wildcard.
